storage/engine/mvcc.go

```diff
@@ -41,38 +40,73 @@ const (
 var (
 	// MVCCKeyMax is a maximum mvcc-encoded key value which sorts after
 	// all other keys.
-	MVCCKeyMax = MVCCEncodeKey(roachpb.KeyMax)
+	MVCCKeyMax = MakeMVCCKey(roachpb.KeyMax)
+	// NilKey is the nil MVCCKey.
+	NilKey = MVCCKey{}
 )
 
 // MVCCKey is an encoded key, distinguished from roachpb.Key in that it
 // is an encoded version.
-type MVCCKey []byte
+type MVCCKey struct {
+	Key       roachpb.Key
+	Timestamp roachpb.Timestamp
+}
+
+// MakeMVCCKey creates an MVCCKey from a roachpb.Key.
+func MakeMVCCKey(key roachpb.Key) MVCCKey {
+	return MVCCKey{Key: key}
+}
 
-// Next returns the next key in lexicographic sort order.
+// Next returns the next key.
 func (k MVCCKey) Next() MVCCKey {
-	return MVCCKey(roachpb.BytesNext(k))
+	ts := k.Timestamp.Prev()
```

Comment:
This panics for keys with zero timestamps, and surprisingly, iterating over `MVCCKey.Next` is not safe because it will eventually advance to a key with a zero timestamp. The doc comments need to be clear about the difference between value and non-value keys and which operations are valid for each. 

Or did you mean to check for `k.Timestamp` being zero before calling `Prev` on it? We'll never see `k.Timestamp.Prev()` returning zero in practice.